### DISCUSSION TITLE
How to migrate with a new required column without default value when there are existing rows?

### DISCUSSION THREAD

**Discussion by maximedupre** [prisma/prisma#20607] [Q&A] [answered]
### Question

I want to add a new required column to my `Experiment` model, but there are existing rows. When I run `prisma migrate dev`, I inevitably get this error message:

> Step 0 Added the required column `website` to the `Experiment` table without a default value. There are 3 rows in this table, it is not possible to execute this step.

That being said, I do not wish to make this column optional or have a default value. Therefore, I must do something about those 3 rows.

Let's say I wanted to add a "dummy" value to those 3 rows or delete those rows (or do any other manipulation that would allow the migration to be applied), where does it fit into the migration? Those steps are part of the migration process but are not necessarily related to a schema modification, yet are needed indirectly to modify the schema.

I appreciate any tips 🙏

**Comment by maximedupre** [comment]
I guess I need to run `prisma migrate dev --create-only` instead, and then manually do the "other manipulation that would allow the migration to be applied" in the generated SQL file.

Since I am new to this, a confirmation that this is the right way to approach this problem would be appreciated 😁🤝

**Reply by ludralph** [reply, marked as answer]
Hi @maximedupre 👋 

I agree that you would need to create a new migration using `prisma migrate dev --create-only`. This will generate a new migration file in your `prisma/migrations` directory. You can then modify the migration script. Open the generated migration SQL script in the migration file. In this script, you'll need to perform the necessary data manipulation to populate the new column with valid data. After modifying the migration script, you can apply the migration using `prisma migrate dev`. 

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by luiz00martins** [reply]
Somewhat unfortunate this is the accepted answer, and there's no way to give a "one-time" default value to be applied during the migration (but not default to the column).

**Reply by gcb** [reply]
here's an example:

before commiting the migration file, edit it:

```psql
-- prisma/migrations/yyyymmdd_migrationname/migration.sql
-- AlterTable
-- original line from prisma migration:
-- ALTER TABLE "public"."xxx" ADD COLUMN     "isEnabled" BOOLEAN NOT NULL DEFAULT false;

-- but we will set all previous one to true,
ALTER TABLE "public"."xxx" ADD COLUMN     "isEnabled" BOOLEAN NOT NULL DEFAULT true;
-- now all legacy rows will have isEnabled=true,
-- and then we set the default false that we want from now on.
ALTER TABLE "public"."xxx" ALTER COLUMN     "isEnabled" SET DEFAULT false;
```

Hope that helps.

**Comment by information-security** [comment]
How is modifying the migration script going to solve the OP's original issue? Let's say we have 1M records and we want to add a unique column without default value.

**Reply by gcb** [reply]
see my comment on the thread above.
